src/providers/aws.rs

```diff
@@ -0,0 +1,858 @@
+    }
+}
+
+impl<P: ProvideAwsCredentials> std::ops::Drop for AWSLauncher<P> {
+    fn drop(&mut self) {}
+}
+
+/// Region specific. Launch AWS EC2 spot instances.
+///
+/// This implementation uses [rusoto](https://crates.io/crates/rusoto_core) to connect to AWS.
+///
+/// EC2 spot instances are normally subject to termination at any point. This library instead
+/// uses [defined duration](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-requests.html#fixed-duration-spot-instances)
+/// instances, which cost slightly more, but are never prematurely terminated.  The lifetime of
+/// such instances must be declared in advance (1-6 hours). By default, we use 6 hours (the
+/// maximum). To change this, AWSRegion respects the limit specified in
+/// [`AWSLauncher::set_max_instance_duration`](AWSLauncher::set_max_instance_duration).
+#[derive(Default)]
+pub struct AWSRegion {
+    pub region: rusoto_core::region::Region,
+    security_group_id: String,
+    ssh_key_name: String,
+    private_key_path: Option<tempfile::NamedTempFile>,
+    client: Option<rusoto_ec2::Ec2Client>,
+    outstanding_spot_request_ids: HashMap<String, (String, MachineSetup)>,
+    instances: HashMap<String, (Option<(String, String)>, (String, MachineSetup))>,
+    log: Option<slog::Logger>,
+}
+
+impl AWSRegion {
+    /// Connect to AWS region `region`, using credentials provider `provider`.
+    ///
+    /// This is a lower-level API, you may want [`AWSLauncher`] instead.
+    ///
+    /// This will create a temporary security group and SSH key in the given AWS region.
+    pub fn new<P>(region: &str, provider: P, log: slog::Logger) -> Result<Self, Error>
+    where
+        P: ProvideAwsCredentials + Send + Sync + 'static,
+        <P as ProvideAwsCredentials>::Future: Send,
+    {
+        let region = region.parse()?;
+        let ec2 = AWSRegion::connect(region, provider, log)?
+            .make_security_group()?
+            .make_ssh_key()?;
+
+        Ok(ec2)
+    }
+
+    fn connect<P>(
+        region: rusoto_core::region::Region,
+        provider: P,
+        log: slog::Logger,
+    ) -> Result<Self, Error>
+    where
+        P: ProvideAwsCredentials + Send + Sync + 'static,
+        <P as ProvideAwsCredentials>::Future: Send,
+    {
+        debug!(log, "connecting to ec2");
+        let ec2 = rusoto_ec2::Ec2Client::new_with(HttpClient::new()?, provider, region.clone());
+
+        Ok(Self {
+            region,
+            security_group_id: Default::default(),
+            ssh_key_name: Default::default(),
+            private_key_path: Some(
+                tempfile::NamedTempFile::new()
+                    .context("failed to create temporary file for keypair")?,
+            ),
+            outstanding_spot_request_ids: Default::default(),
+            instances: Default::default(),
+            client: Some(ec2),
+            log: Some(log),
+        })
+    }
+
+    fn make_security_group(mut self) -> Result<Self, Error> {
```

Comment:
new() needs to return Self, and internally it calls two methods that consume and return Self.
These could be made `&mut Self`, but that makes the code in `new()` ugly, something like: 

```
let ec2 = AWSRegion::connect(...)?;
ec2.make_security_group(..)?.make_ssh_key(..)?;
Ok(ec2)
```

rather that the current one-liner.